doc/designs/lms.md

```diff
@@ -0,0 +1,243 @@
+  added if this is ever required.
+- CMS and X509 support - this may be added at a later date.
+- Implementing Extended Merkle signature scheme (XMSS) is a separate effort.
+
+HSS Signature and Key formats
+-----------------------------
+
+The LMS/HSS signature and public key formats use XDR (not ASN1 data).
+The reference code in RFC 8554 checks sizes of the binary data before loading
+the known fields. Rather than do this the OpenSSL PACKET interface is used.
+When the end of the data is expected PACKET_remaining() is tested.
+NOTE that the size of signatures depends on the modes which are embedded in the
+loaded data.
+Byte array data such as encoded public keys are just kept as pointers into the
+data. The entire signature blob is duplicated if the pointers are required
+across API calls.
+
+HSS message streaming API
+-------------------------
+
+In order to handle HSS message streaming the signature data must be decoded
+first, since the message is only used in the last part of the verify.
+This means that the existing sequence of calls for streaming (i.e.
+DigestVerifyInit/DigestVerifyUpdate/DigestFinal) is not sufficient for this
+purpose unless the message is buffered during the update, and then only
+processed during the final.
+
+So we should:
+
+- Support one shot operations:
+  Using existing EVP_DigestVerifyInit_ex + EVP_DigestVerify calls
+- Dont support streaming via existing interface:
+  EVP_DigestVerifyInit_ex + EVP_DigestVerifyUpdate + EVP_DigestVerifyFinal.
+  It could be done, but requires buffering of the message during updates.
+- EVP_VerifyInit etc should not be supported.
+- In order to set the signature early one of the following methods should be
+  used:
+
+Proposal 1:
+
+Add streaming via a new API that is similar to the existing DigestVerify API's,
+but moves the signature |sig| from the final to the init. i.e.
+
+```c
+__owur int EVP_DigestVerifyHBSInit(EVP_MD_CTX *ctx, EVP_PKEY_CTX **pctx,
+                                   const char *mdname,
+                                   OSSL_LIB_CTX *libctx,
+                                   const char *props, EVP_PKEY *pkey,
+                                   const OSSL_PARAM params[],
+                                   const unsigned char *sig,
+                                   size_t siglen);   
+/* Alias: since the function signature is identical */
+#define EVP_DigestVerifyHBSUpdate(a,b,c) EVP_DigestVerifyUpdate(a,b,c)
+
+__owur int EVP_DigestVerifyHBSFinal(EVP_MD_CTX *ctx);
+```
+
+Proposal 2:
+
+Add a Init_ex function and leave the Final the same.
+
+i.e.
+EVP_DigestVerifyInit_ex3(..., const unsigned char *sig, size_t siglen);
+
+For the EVP_DigestVerifyFinal pass a NULL signature.
+The signature length must be zero in this case or an error will occur.
+If a signature is passed should it just ignore the value?
+
+Proposal 3:
+
+Pass the signature after the init via an OSSL_PARAM.
+
+
```

Comment:
We talked about both of theses type of schemes in PKCS #11: 1) adding just a new init function that takes the signature and 2) adding the signature to the mechanism parameter (the equivalent of OSSL_PARAM). Both of these generated the question of what do you return on Final if the signature is included; a) failure if it's not null, b) failure if it's not null and doesn't match, or c) undefined if the application includes the signature. In the end having a complete new set of C_VerifySignatureXXXX functions got rid of all those issues since the Final versions won't have a signature. We made the C_VerifySignatureUpdate a separate function, but your proposal to make it a #define for the normal update is actually good (so applications that need to support both don't have to keep track of which to call the correct update function.